### DISCUSSION TITLE
Nest.js, Typescript, Prisma and nested objects

### DISCUSSION THREAD

**Discussion by mjstrobl** [prisma/prisma#10729] [Q&A]
Hi,

We're having issues with Nest.js, Typescript,  Prisma and nested objects. We hope to find some help here.

We are using a Postgres database with four relations: Area, User, Objective and UsersOnObjectives. User and Objective have an n:m relationship, explicitly modelled with UsersOnObjectives. Objective has the object "teams" for users and when a new objective is created, we would like to include an array of users and add it to UsersOnObjectives.

Here's a our schema:

```
model Area {
  id       String  @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
}

model User {
  id          String  @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
}

model Objective {
  id       String  @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
  area       Area        @relation(fields: [areaId], references: [id])
  areaId     String      @db.Uuid
  team UsersOnObjectives[]
}

model UsersOnObjectives {
  user        User      @relation(fields: [userId], references: [id])
  userId      String    @db.Uuid
  objective   Objective @relation(fields: [objectiveId], references: [id])
  objectiveId String    @db.Uuid

  @@id([userId, objectiveId])
}

```

Here's the DTO for creating objectives:

```
export class CreateObjectiveDto {
  @ApiProperty()
  @IsNotEmpty()
  areaId: string

  @ApiPropertyOptional()
  @IsOptional()
  @IsDefined()
  @Type(() => CreateUserOnObjectivesDto)
  @ValidateNested({ each: true })
  team?: UsersOnObjectives[]
}
```

Create new objectives (controller and then service):

```
create(@Body() data: CreateObjectiveDto): Promise<Objective> {
    return this.objectiveService.create({ ...data })
  }

async create(payload: Prisma.ObjectiveCreateInput): Promise<Objective> {   
    return await this.prisma.objective.create({data:payload})
}
```

Prisma.ObjectiveCreateInput declaration:

```
export type ObjectiveCreateInput = {
    id?: string
    area?: AreaCreateNestedOneWithoutObjectivesInput
    team?: UsersOnObjectivesCreateNestedManyWithoutObjectiveInput
  }
```

There are two issues here:
1. ObjectiveCreateInput contains "area", not just "areaId". Therefore "create()" (independent from whether a team is included) does not work since an area object is always expected, although the DTO only contains the "areaId". I expected that the  database schema already defines how to resolve "areaId" and including the area object is not necessary. I see why the error happens, but I don't understand why ObjectiveCreateInput looks like that. Do you guys know if we are missing something here? Unfortunately I can't figure how to fix this, or what the proper types are to be used here.
2. For the nested object "team" in Objective, the expected type is UsersOnObjectivesCreateNestedManyWithoutObjectiveInput, although it should be UsersOnObjectives[] from the DTO. This results in the following error when an objective including a team is expected to be created: "Provided List\<Json\>, expected UsersOnObjectivesUncheckedCreateNestedManyWithoutObjectiveInput" The Frontend sends an array of users as "team" as json objects. Ideally this results in a list of users or UsersOnObjectives, so we'd like to create the objective and the team all at once. I actually don't quite understand how this should work, although from the documentation I thought it may be possible.

I guess we're missing a few things here and we appreciate any help to figure out what the issues are with our code.

Thanks.

**Comment by AlejandroGB13** [comment]
Hey, have you solved the problem?. We're facing the same one.

**Reply by pstark-asco** [reply]
Hello there.
Facing the same problem. How did you manage it?
